llvm/lib/Target/AMDGPU/SIRegisterInfo.cpp

```diff
@@ -2523,22 +2532,70 @@ bool SIRegisterInfo::eliminateFrameIndex(MachineBasicBlock::iterator MI,
 
             // We may have 1 free scratch SGPR even though a carry out is
             // unavailable. Only one additional mov is needed.
-            Register TmpScaledReg = RS->scavengeRegisterBackwards(
-                AMDGPU::SReg_32_XM0RegClass, MI, false, 0, false);
-            Register ScaledReg = TmpScaledReg.isValid() ? TmpScaledReg : FrameReg;
+            Register TmpScaledReg = IsCopy && IsSALU
+                                        ? ResultReg
+                                        : RS->scavengeRegisterBackwards(
+                                              AMDGPU::SReg_32_XM0RegClass, MI,
+                                              false, 0, /*AllowSpill=*/false);
+            Register ScaledReg =
+                TmpScaledReg.isValid() ? TmpScaledReg : FrameReg;
+            Register TmpResultReg = ScaledReg;
+
+            if (!LiveSCC) {
+              BuildMI(*MBB, MI, DL, TII->get(AMDGPU::S_LSHR_B32), TmpResultReg)
+                  .addReg(FrameReg)
+                  .addImm(ST.getWavefrontSizeLog2());
+              BuildMI(*MBB, MI, DL, TII->get(AMDGPU::S_ADD_I32), TmpResultReg)
+                  .addReg(TmpResultReg, RegState::Kill)
+                  .addImm(Offset);
+            } else {
+              TmpResultReg = RS->scavengeRegisterBackwards(
+                  AMDGPU::VGPR_32RegClass, MI, false, 0, /*AllowSpill=*/true);
+              BuildMI(*MBB, MI, DL, TII->get(AMDGPU::V_LSHR_B32_e64),
+                      TmpResultReg)
+                  .addImm(ST.getWavefrontSizeLog2())
+                  .addReg(FrameReg);
+
+              MachineInstrBuilder Add;
+              if ((Add = TII->getAddNoCarry(*MBB, MI, DL, TmpResultReg, *RS)) !=
+                  nullptr) {
+                if (Add->getOpcode() == AMDGPU::V_ADD_CO_U32_e64) {
+                  BuildMI(*MBB, *Add, DL, TII->get(AMDGPU::S_MOV_B32),
+                          ResultReg)
+                      .addImm(Offset);
+                  Add.addReg(ResultReg, RegState::Kill)
+                      .addReg(TmpResultReg, RegState::Kill)
+                      .addImm(0);
+                } else
+                  Add.addImm(Offset).addReg(TmpResultReg, RegState::Kill);
+              } else {
+                // VCC is live and no SGPR is free.
+                // since emergency stack slot is already used for spilling VGPR
+                // scavenged? This a way around to avoid carry, need follow-up.
+                BuildMI(*MBB, MI, DL, TII->get(AMDGPU::S_MOV_B32), ResultReg)
+                    .addImm(Offset);
```

Comment:
Instead you can materialize in the VGPR TmpResultReg. That saves the constant bus use for the frame pointer register to directly materialize in one mad

Reply:
It will clobber shift value right?

Reply:
Not if you replace the whole sequence. It's the free register you just scavenged, so it would be 
tmp_vgpr = offset
tmp_vgpr = mad24 FrameReg, wavesize, tmp_vgpr

Reply:
Also you should add an assert that the offset is positive. I think we need to switch this to the signed mad when we switch the scratch instruction ABI to use negative offsets

Reply:
> Not if you replace the whole sequence. It's the free register you just scavenged, so it would be tmp_vgpr = offset tmp_vgpr = mad24 FrameReg, wavesize, tmp_vgpr

That's really great idea.

Reply:
> Also you should add an assert that the offset is positive. I think we need to switch this to the signed mad when we switch the scratch instruction ABI to use negative offsets

sure, will add it.

Reply:
We start with a frame pointer with a wave space value, and an offset in lane-space. We are materializing a lane space value. You can either do a right shift of the frame pointer to get to lane space, or a left shift of the offset to get to wavespace. We can right shift after the computation to get back to the desired per-lane value.

I also spent some time with alive2, and unfortunately we can only use the mad24 on gfx9 and gfx10. https://alive2.llvm.org/ce/z/h8Zkdw. The stack size increased in gfx11 so that the maximum sized object computation can no longer be done with the 24-bit multiply.

Reply:
Actually we might be able to - I wrote the assumes, assuming the FP and the offset are the maximum. But these are actually linked. The condition is more like fp + offset < maximum value

Reply:
Hmm. Still seems like it would require a 26-bit multiply in wave32

Reply:
I have made changes, is it what you have suggested.